File: apps/lark-server/src/infrastructure/integrations/rabbitmq.ts
import amqplib, { Channel, ChannelModel, ConsumeMessage } from 'amqplib';

const EXCHANGE_NAME = 'post_processing';
const DLX_NAME = 'post_processing_dlx';
const DLQ_NAME = 'dead_letters';

export const QUEUE_RECALL = 'recall';
export const QUEUE_VECTORIZE = 'vectorize';
export const QUEUE_CHAT_REQUEST = 'chat_request';
export const QUEUE_CHAT_RESPONSE = 'chat_response';
export const RK_RECALL = 'action.recall';
export const RK_VECTORIZE = 'task.vectorize';
export const RK_CHAT_REQUEST = 'chat.request';
export const RK_CHAT_RESPONSE = 'chat.response';

const NON_PROD_EXPIRES_MS = 86_400_000;
const LANE_FALLBACK_TTL_MS = 10_000;

export type MessageHandler = (msg: ConsumeMessage) => Promise<void>;

/** 获取当前泳道：读环境变量，prod/空返回 undefined */
export function getLane(): string | undefined {
    const lane = process.env.LANE;
    if (!lane || lane === 'prod') return undefined;
    return lane;
}

/** 泳道队列名：base 或 base_{lane} */
export function laneQueue(base: string, lane?: string): string {
    return lane ? `${base}_${lane}` : base;
}

/** 泳道 routing key：base 或 base.{lane} */
export function laneRK(base: string, lane?: string): string {
    return lane ? `${base}.${lane}` : base;
}

class RabbitMQClient {
    private static instance: RabbitMQClient;
    private conn: ChannelModel | null = null;
    private channel: Channel | null = null;
    private reconnecting = false;
    private consumers: Array<{ queue: string; handler: MessageHandler }> = [];

    private constructor() {}

    static getInstance(): RabbitMQClient {
        if (!RabbitMQClient.instance) {
            RabbitMQClient.instance = new RabbitMQClient();
        }
        return RabbitMQClient.instance;
    }

    async connect(): Promise<void> {
        if (this.channel) return;

        const url = process.env.RABBITMQ_URL;
        if (!url) {
            throw new Error('RABBITMQ_URL is not configured');
        }

        this.conn = await amqplib.connect(url);

        this.conn.on('error', (err: Error) => {
            console.error('[RabbitMQ] connection error:', err.message);
        });
        this.conn.on('close', () => {
            console.warn('[RabbitMQ] connection closed, will reconnect');
            this.channel = null;
            this.conn = null;
            this.scheduleReconnect();
        });

        this.channel = await this.conn.createChannel();
        await this.channel.prefetch(10);
        console.info('[RabbitMQ] connected');
    }

    async declareTopology(): Promise<void> {
        const ch = this.getChannel();
        const lane = getLane();

        // DLX + DLQ
        await ch.assertExchange(DLX_NAME, 'fanout', { durable: true });
        await ch.assertQueue(DLQ_NAME, { durable: true });
        await ch.bindQueue(DLQ_NAME, DLX_NAME, '');

        // Main exchange (delayed-message)
        await ch.assertExchange(EXCHANGE_NAME, 'x-delayed-message', {
            durable: true,
            arguments: { 'x-delayed-type': 'topic' },
        });

        // 队列参数：prod 队列用 DLX → DLQ；lane 队列用 TTL → 主 exchange fallback 到 prod
        function queueArgs(prodRK: string): Record<string, unknown> {
            const extra: Record<string, unknown> = lane
                ? { 'x-expires': NON_PROD_EXPIRES_MS }
                : {};
            if (!lane) {
                return { 'x-dead-letter-exchange': DLX_NAME, ...extra };
            }
            return {
                'x-message-ttl': LANE_FALLBACK_TTL_MS,
                'x-dead-letter-exchange': EXCHANGE_NAME,
                'x-dead-letter-routing-key': prodRK,
                ...extra,
            };
        }

        // recall queue
        const recallQ = laneQueue(QUEUE_RECALL, lane);
        await ch.assertQueue(recallQ, { durable: true, arguments: queueArgs(RK_RECALL) });
        await ch.bindQueue(recallQ, EXCHANGE_NAME, laneRK(RK_RECALL, lane));

        // vectorize queue
        const vectorizeQ = laneQueue(QUEUE_VECTORIZE, lane);
        await ch.assertQueue(vectorizeQ, { durable: true, arguments: queueArgs(RK_VECTORIZE) });
        await ch.bindQueue(vectorizeQ, EXCHANGE_NAME, laneRK(RK_VECTORIZE, lane));

        // chat_request queue
        const chatReqQ = laneQueue(QUEUE_CHAT_REQUEST, lane);
        await ch.assertQueue(chatReqQ, { durable: true, arguments: queueArgs(RK_CHAT_REQUEST) });
        await ch.bindQueue(chatReqQ, EXCHANGE_NAME, laneRK(RK_CHAT_REQUEST, lane));

        // chat_response queue
        const chatRespQ = laneQueue(QUEUE_CHAT_RESPONSE, lane);
        await ch.assertQueue(chatRespQ, { durable: true, arguments: queueArgs(RK_CHAT_RESPONSE) });
        await ch.bindQueue(chatRespQ, EXCHANGE_NAME, laneRK(RK_CHAT_RESPONSE, lane));

        console.info(`[RabbitMQ] topology declared (lane=${lane || 'prod'})`);
    }

    async publish(
        routingKey: string,
        body: Record<string, unknown>,
        delayMs?: number,
        headers?: Record<string, unknown>,
        lane?: string,
    ): Promise<void> {
        const ch = this.getChannel();

        // 默认取当前泳道；传入 lane 则使用传入值
        const effectiveLane =
            lane !== undefined ? (lane === 'prod' ? undefined : lane) : getLane();
        const actualRK = laneRK(routingKey, effectiveLane);

        const msgHeaders: Record<string, unknown> = { ...headers };
        if (delayMs !== undefined) {
            msgHeaders['x-delay'] = delayMs;
        }

        ch.publish(EXCHANGE_NAME, actualRK, Buffer.from(JSON.stringify(body)), {
            persistent: true,
            contentType: 'application/json',
            headers: Object.keys(msgHeaders).length > 0 ? msgHeaders : undefined,
        });
    }

    async consume(queueName: string, handler: MessageHandler): Promise<void> {
        // 记录 consumer 以便重连后恢复
        if (!this.consumers.some((c) => c.queue === queueName)) {
            this.consumers.push({ queue: queueName, handler });
        }
        await this.registerConsumer(queueName, handler);
    }

    private async registerConsumer(queueName: string, handler: MessageHandler): Promise<void> {
        const ch = this.getChannel();
        await ch.consume(queueName, async (msg) => {
            if (!msg) return;
            try {
                await handler(msg);
            } catch (err) {
                console.error(`[RabbitMQ] handler error on ${queueName}:`, err);
                ch.nack(msg, false, false);
            }
        });
        console.info(`[RabbitMQ] consuming queue: ${queueName}`);
    }

    ack(msg: ConsumeMessage): void {
        this.getChannel().ack(msg);
    }

    nack(msg: ConsumeMessage, requeue = false): void {
        this.getChannel().nack(msg, false, requeue);
    }

    getChannel(): Channel {
        if (!this.channel) {
            throw new Error('RabbitMQ channel not available; call connect() first');
        }
        return this.channel;
    }

    async close(): Promise<void> {
        try {
            await this.channel?.close();
            await this.conn?.close();
        } catch {
            // ignore close errors
        }
        this.channel = null;
        this.conn = null;
    }

    private scheduleReconnect(): void {
        if (this.reconnecting) return;
        this.reconnecting = true;
        setTimeout(async () => {
            this.reconnecting = false;
            try {
                await this.connect();
                await this.declareTopology();
                for (const { queue, handler } of this.consumers) {
                    await this.registerConsumer(queue, handler);
                }
                console.info('[RabbitMQ] reconnected');
            } catch (err) {
                console.error('[RabbitMQ] reconnect failed:', err);
                this.scheduleReconnect();
            }
        }, 5000);
    }
}

export const rabbitmqClient = RabbitMQClient.getInstance();

File: apps/agent-service/app/clients/rabbitmq.py
"""RabbitMQ 客户端 — 单例，声明拓扑 + 发布/消费"""

import json
import logging
import os
from collections.abc import Callable, Coroutine
from typing import Any, NamedTuple

import aio_pika
from aio_pika import DeliveryMode, ExchangeType, Message
from aio_pika.abc import AbstractIncomingMessage

from app.config import settings

logger = logging.getLogger(__name__)

# 拓扑常量
EXCHANGE_NAME = "post_processing"
DLX_NAME = "post_processing_dlx"
DLQ_NAME = "dead_letters"


# Route 当前假设 queue:rk = 1:1，未来如果需要演进为更灵活的结构需另行讨论。
class Route(NamedTuple):
    queue: str
    rk: str


CHAT_REQUEST = Route("chat_request", "chat.request")
CHAT_RESPONSE = Route("chat_response", "chat.response")
SAFETY_CHECK = Route("safety_check", "post.safety.check")
RECALL = Route("recall", "action.recall")
VECTORIZE = Route("vectorize", "task.vectorize")

ALL_ROUTES = [CHAT_REQUEST, CHAT_RESPONSE, SAFETY_CHECK, RECALL, VECTORIZE]

# 非 prod 队列空闲自动删除（24h）
_NON_PROD_EXPIRES_MS = 86_400_000
_LANE_FALLBACK_TTL_MS = 10_000


def _current_lane() -> str | None:
    """获取当前泳道：优先 contextvars，fallback 到环境变量"""
    try:
        from app.utils.middlewares.trace import get_lane

        lane = get_lane()
    except Exception:
        lane = None
    if not lane:
        lane = os.getenv("LANE")
    if not lane or lane == "prod":
        return None
    return lane


def _lane_queue(base: str, lane: str | None) -> str:
    """返回泳道队列名：base 或 base_{lane}"""
    return f"{base}_{lane}" if lane else base


def _lane_rk(base: str, lane: str | None) -> str:
    """返回泳道 routing key：base 或 base.{lane}"""
    return f"{base}.{lane}" if lane else base

def _build_queue_args(prod_rk: str, lane: str | None) -> dict[str, Any]:
    """构建队列参数：prod 队列用 DLX → DLQ；lane 队列用 TTL → 主 exchange fallback 到 prod"""
    extra: dict[str, Any] = {}
    if lane:
        extra["x-expires"] = _NON_PROD_EXPIRES_MS
    if not lane:
        return {"x-dead-letter-exchange": DLX_NAME, **extra}
    return {
        "x-message-ttl": _LANE_FALLBACK_TTL_MS,
        "x-dead-letter-exchange": EXCHANGE_NAME,
        "x-dead-letter-routing-key": prod_rk,
        **extra,
    }


MessageHandler = Callable[[AbstractIncomingMessage], Coroutine[Any, Any, None]]


class RabbitMQClient:
    """aio-pika 单例客户端"""

    _instance: "RabbitMQClient | None" = None

    def __init__(self) -> None:
        self._connection: aio_pika.abc.AbstractRobustConnection | None = None
        self._channel: aio_pika.abc.AbstractRobustChannel | None = None
        self._exchange: aio_pika.abc.AbstractExchange | None = None
        self._declared_lane_queues: set[str] = set()

    @classmethod
    def get_instance(cls) -> "RabbitMQClient":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    async def connect(self) -> None:
        if self._connection and not self._connection.is_closed:
            return
        url = settings.rabbitmq_url
        if not url:
            raise RuntimeError("RABBITMQ_URL is not configured")
        self._connection = await aio_pika.connect_robust(url)
        self._channel = await self._connection.channel()
        await self._channel.set_qos(prefetch_count=10)
        self._declared_lane_queues = set()
        logger.info("RabbitMQ connected: %s", url.split("@")[-1])

    async def declare_topology(self) -> None:
        """声明 exchange、queue、binding、DLX（按泳道隔离）"""
        assert self._channel is not None, "must call connect() first"

        lane = _current_lane()

        # DLX + DLQ
        dlx = await self._channel.declare_exchange(
            DLX_NAME, ExchangeType.FANOUT, durable=True
        )
        dlq = await self._channel.declare_queue(DLQ_NAME, durable=True)
        await dlq.bind(dlx)

        # 主 exchange (delayed-message)
        self._exchange = await self._channel.declare_exchange(
            EXCHANGE_NAME,
            type="x-delayed-message",
            durable=True,
            arguments={"x-delayed-type": "topic"},
        )

        for route in ALL_ROUTES:
            q = await self._channel.declare_queue(
                _lane_queue(route.queue, lane),
                durable=True,
                arguments=_build_queue_args(route.rk, lane),
            )
            await q.bind(
                self._exchange, routing_key=_lane_rk(route.rk, lane)
            )

        logger.info("RabbitMQ topology declared (lane=%s)", lane or "prod")

    async def _ensure_lane_queue(self, route: Route, lane: str) -> None:
        """懒声明泳道队列：publish 时若目标泳道队列尚未声明，则自动创建并绑定"""
        cache_key = f"{route.queue}_{lane}"
        if cache_key in self._declared_lane_queues:
            return
        assert self._channel is not None, "must call connect() first"
        q = await self._channel.declare_queue(
            _lane_queue(route.queue, lane),
            durable=True,
            arguments=_build_queue_args(route.rk, lane),
        )
        await q.bind(self._exchange, routing_key=_lane_rk(route.rk, lane))
        self._declared_lane_queues.add(cache_key)
        logger.info("Lazy-declared lane queue: %s_%s", route.queue, lane)

    async def publish(
        self,
        route: Route,
        body: dict,
        delay_ms: int | None = None,
        headers: dict | None = None,
        lane: str | None = ...,  # type: ignore[assignment]
    ) -> None:
        """发布消息。lane 默认取 _current_lane()，传 None 强制 prod。"""
        assert self._exchange is not None, "must call declare_topology() first"

        if lane is ...:
            lane = _current_lane()
        # prod / 空 → None
        if lane == "prod":
            lane = None

        if lane:
            await self._ensure_lane_queue(route, lane)

        actual_rk = _lane_rk(route.rk, lane)

        msg_headers: dict[str, Any] = headers or {}
        if delay_ms is not None:
            msg_headers["x-delay"] = delay_ms

        message = Message(
            body=json.dumps(body).encode(),
            delivery_mode=DeliveryMode.PERSISTENT,
            content_type="application/json",
            headers=msg_headers if msg_headers else None,
        )
        await self._exchange.publish(message, routing_key=actual_rk)

    async def consume(self, queue_name: str, callback: MessageHandler) -> None:
        assert self._channel is not None, "must call connect() first"

        queue = await self._channel.get_queue(queue_name)
        await queue.consume(callback)
        logger.info("Consuming queue: %s", queue_name)

    async def close(self) -> None:
        if self._connection and not self._connection.is_closed:
            await self._connection.close()
            logger.info("RabbitMQ connection closed")
